### ISSUE TITLE
Shuffle between workers before returning batch?

### ISSUE THREAD

**Issue by jopo666** [webdataset/webdataset#141]
Hi!

Currently `webdataset` works perfectly for me with individual datasets, but I have several different datasets that I would like to combine like this.

```
urls_1 = ["./dogs_1.tar", "./dogs_2.tar", ..., "/dogs_n.tar"]
urls_2 = ["./cats_1.tar", "./cats_2.tar", ..., "/cats_n.tar"]
```

Then I combine (and shuffled) the `urls` and pass them to `WebDataset` and `DataLoader`. Now each worker receives a set of shards and shuffles between them. No matter how high the `shuffle_buffer` is the batch will contain samples from only these shards. Here lies the problem where the shuffling isn't working anymore. 

Let's say `__worker__ [1/32]` received 4 dog shards and 4 cat shards. Then the distribution of dogs and cats in a single batch would be close to 50%. But due to shard shuffling it is likely that some workers receive 2 dog and 6 cat shards, and then single batch cat/dog distribution would be way off from the actual distribution. The problem is more severe if I have a third set of urls eg. `["dragons_1.tar", "dragons_2.tar"]`, which are really rare.

Is there a way (with `webdataset`) to shuffle different worker outputs to remove the problem? I'm guessing I could also create another shuffle buffer inside the training loop, for example....

```python
shuffle_queue = FancyBatchShuffler3000(shuffle_buffer=1000)

for worker_batch in loader:
    # Add batch to shuffle queue.
    shuffle_queue.add_batch(worker_batch)
    # Get batch which contains samples from many workers.
    batch = shuffle_queue.get_shuffled_batch()
```

... but I'd rather handle everything inside `webdataset`.

**Comment by tmbdev** [comment]
Yes: to shuffle between workers, use WebDataloader (it's just a thin wrapper) and rebatch:

```
dataset = WebDataset(...).batched(ds_batch_size)
dataloader = WebDataloader(dataset, ...).unbatched().shuffle(10000).batched(batch_size)
```

This has the added advantage of making it easy to change the batch size on the fly.

The `ds_batch_size` is chosen for optimizing communication between dataset and dataloader and doesn't depend on your GPU memory anymore (it depends on your system, but my guess is that a `ds_batch_size` resulting in transfers of a few megabytes is good).
